server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java

```diff
@@ -12,69 +12,71 @@
 import java.util.Objects;
-import java.util.concurrent.ConcurrentMap;
 
 /** Utility class to resolve the Lucene doc ID, version, seqNo and primaryTerms for a given uid. */
 public final class VersionsAndSeqNoResolver {
 
-    static final ConcurrentMap<IndexReader.CacheKey, CloseableThreadLocal<PerThreadIDVersionAndSeqNoLookup[]>> lookupStates =
-        ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+    private static final PerThreadIDVersionAndSeqNoLookup[] EMPTY_LOOKUP = new PerThreadIDVersionAndSeqNoLookup[0];
 
-    // Evict this reader from lookupStates once it's closed:
-    private static final IndexReader.ClosedListener removeLookupState = key -> {
-        CloseableThreadLocal<PerThreadIDVersionAndSeqNoLookup[]> ctl = lookupStates.remove(key);
-        if (ctl != null) {
-            ctl.close();
-        }
-    };
+    private static final ThreadLocal<PerThreadVersionsAndSeqNoResolver> resolversByThread = ThreadLocal.withInitial(
+        PerThreadVersionsAndSeqNoResolver::new
+    );
 
-    private static PerThreadIDVersionAndSeqNoLookup[] getLookupState(IndexReader reader, String uidField) throws IOException {
-        // We cache on the top level
+    private static class PerThreadVersionsAndSeqNoResolver {
+        // We cache on the top level.
         // This means cache entries have a shorter lifetime, maybe as low as 1s with the
         // default refresh interval and a steady indexing rate, but on the other hand it
         // proved to be cheaper than having to perform a CHM and a TL get for every segment.
         // See https://github.com/elastic/elasticsearch/pull/19856.
-        IndexReader.CacheHelper cacheHelper = reader.getReaderCacheHelper();
-        CloseableThreadLocal<PerThreadIDVersionAndSeqNoLookup[]> ctl = lookupStates.get(cacheHelper.getKey());
-        if (ctl == null) {
-            // First time we are seeing this reader's core; make a new CTL:
-            ctl = new CloseableThreadLocal<>();
-            CloseableThreadLocal<PerThreadIDVersionAndSeqNoLookup[]> other = lookupStates.putIfAbsent(cacheHelper.getKey(), ctl);
-            if (other == null) {
-                // Our CTL won, we must remove it when the reader is closed:
-                cacheHelper.addClosedListener(removeLookupState);
-            } else {
-                // Another thread beat us to it: just use their CTL:
-                ctl = other;
-            }
-        }
 
-        PerThreadIDVersionAndSeqNoLookup[] lookupState = ctl.get();
-        if (lookupState == null) {
-            lookupState = new PerThreadIDVersionAndSeqNoLookup[reader.leaves().size()];
-            for (LeafReaderContext leaf : reader.leaves()) {
-                lookupState[leaf.ord] = new PerThreadIDVersionAndSeqNoLookup(leaf.reader(), uidField);
-            }
-            ctl.set(lookupState);
-        }
+        private final Map<IndexReader.CacheKey, PerThreadIDVersionAndSeqNoLookup[]> lookupsByReader = ConcurrentCollections
```

Comment:
Not sure about using a CHM here, nor about its exact config. We only put and read on the owning thread, the thread-safety is just needed for the cleanup operations which should be much less common. Maybe a synchronized map would be fine/better?